CMakeLists.txt

```diff
@@ -70,6 +72,43 @@ macro(apply_llvm_compile_flags target)
     target_compile_definitions(${target} PRIVATE ${LLVM_DEFINITIONS})
 endmacro()
 
+# Check if using external LLVM/MLIR install tree
+if (DEFINED MLIR_DIR)
+    if (DEFINED MLIR_DIR)
+        set(MLIR_EXT_USE_LLVM_INSTALL_TREE ON)
+    else()
+        message(FATAL_ERROR "Need to define both LLVM_DIR and MLIR_DIR or none.")
+    endif()
+elseif (DEFINED LLVM_DIR)
+    message(FATAL_ERROR "Need to define both LLVM_DIR and MLIR_DIR or none.")
+else()
+    set(MLIR_EXT_USE_LLVM_INSTALL_TREE OFF)
+endif()
+
+if(GPU_ENABLE)
+    # Check if using external level-zero
+    if(DEFINED LEVEL_ZERO_DIR)
+        set(MLIR_EXT_USE_EXT_LEVEL_ZERO ON)
+    else()
+        set(MLIR_EXT_USE_EXT_LEVEL_ZERO OFF)
+    endif()
+    message(STATUS "MLIR_EXT_USE_EXT_LEVEL_ZERO: ${MLIR_EXT_USE_EXT_LEVEL_ZERO}")
+endif()
+
+add_subdirectory(third-party)
+
+if(NOT MLIR_EXT_USE_LLVM_INSTALL_TREE)
+    set(LLVM_DIR ${PROJECT_BINARY_DIR}/third-party/mlir/llvm-project/llvm/lib/cmake/llvm)
```

Comment:
I do not like checking out third party repos in IMEX repo. The problem is if you use `git clean -dfx` the folder will be wiped out. In my Python build_locally.py, I allow users to set a `WORKING_DIR` or use system temp directory. Let us at least have the option of providing a -DCMAKE_THIRD_PARTY_LIB_PATH` or something like that?

Reply:
The PR does not remove the option of providing -DCMAKE_THIRD_PARTY_LIB_PATH. It only checks out third party repo if -DCMAKE_THIRD_PARTY_LIB_PATH is not provided.